### PR TITLE
Dgad 862: optional password complexity rules for apostrophe

### PR DESCRIPTION
**Pull request by boutell**
`passwordMinLength`

The minimum length for passwords. You should set this, as there
is no default for bc reasons (effectively the default is `1`).

`passwordRules`

An optional array of password rule names, as strings. The standard rules
available are `noSlashes`, `noSpaces`, `mixedCase`, `digits`, and
`noTripleRepeats`. The `noTripleRepeats` rule forbids repeating a
character three times in a row. By default no rules are in effect.

When this option is set, the rules are consulted when a password
is set or reset. Existing passwords that do not follow the rules
are tolerated. If you wish to enforce them for existing passwords
as well, see below.

`passwordRulesAtLoginTime`

By default, password rules are enforced only when a password is 
being set or reset. If you wish, you can set `passwordRulesAtLoginTime: true`
to apply them at login time so that even an existing password will
not work unless it passes the rules. This can be useful if you don't
mind a few irritated users and you have the `passwordReset` option
properly configured (see below).

### CODE REVIEW

**Comment by boutell** [comment]
Sample config in `data/local.js`:

```
module.exports = {
  modules: {
    'apostrophe-login': {
      passwordMinLength: 10,
      passwordRules: [ 'noSlashes', 'noSpaces', 'mixedCase', 'digits', 'noTripleRepeats' ],
      passwordRulesAtLoginTime: true
    }
  }
};
```

**Comment by boutell** [comment]
Other new things of note in this PR:

* The `editor-modal.js` of pieces now has a displayError method invoked after a server-side save error, which, by default, does the existing thing: display a notification. This is a useful point at which to override.

* In `apostrophe-users`, an override of this creates and displays the full set of password rule errors when needed.

* To facilitate passing an array of error messages up from deeply nested code without a big bc break in how errors are propagated in Apostrophe, in the event of an error, the contents of `req.errorMessages` are now included in the `messages` property of the response from the APIs that save pieces. `req.errorMessages` can be populated anywhere, even if a simple string error must be passed up from that point for compatibility reasons.

* Password schema fields always use `autocomplete="off"`. Note this does not interfere with login, which does not use schema based forms. It's to prevent weirdness when editing an account as an admin, etc.

Care has been taken to preserve bc, even with existing template overrides.

File: lib/modules/apostrophe-login/index.js [hunk 1]
```diff
@@ -315,22 +368,52 @@ module.exports = {
...
 
         self.apos.app.post('/password-reset', function(req, res) {
           var reset = self.apos.launder.string(req.body.reset);
           var email = self.apos.launder.string(req.body.email);
           var password = self.apos.launder.string(req.body.password);
-          if ((!reset.length) || (!password.length)) {
-            return res.redirect('/password-reset-request?error=missing');
+          if (!reset.length) {
+            return res.redirect('/password-reset?' + qs.stringify({
+              errors: [ 'The password reset code is missing from your link. Try resetting your password again.' ],
```
Comment (abea) [inline lib/modules/apostrophe-login/index.js]: Suggestion: "The password reset code is missing from your URL. Click the link in your reset email or try resetting your password again."

File: lib/modules/apostrophe-login/index.js [hunk 2]
```diff
@@ -315,22 +368,52 @@ module.exports = {
...
             return res.redirect('/login?message=' + req.__('That reset code was not found. It may have expired. Try resetting again.'));
           });
+
+          function getMessage() {
+            let message;
+            if (req.query.message) {
+              message = req.query.message;
+            }
+            if (req.query.errors && req.query.errors.length) {
+              message = req.query.errors.join(', ');
```
Comment (abea) [inline lib/modules/apostrophe-login/index.js]: This seems like it'll read weirdly. Why not write the messages as sentences and join with a space?

File: lib/modules/apostrophe-login/index.js [hunk 3]
```diff
@@ -454,6 +537,89 @@ module.exports = {
...
+      });
+      return errors;
+    };
+
+    self.passwordRules = {
+      noSlashes: {
+        test: function(password) {
+          return !((password.indexOf('/') !== -1) || (password.indexOf('\\') !== -1));
+        },
+        message: '/ and \\ characters are not allowed in password'
```
Comment (abea) [inline lib/modules/apostrophe-login/index.js]: I guess the assumption with these is that they would be overridden for non-English?

Comment (boutell) [reply lib/modules/apostrophe-login/index.js]: No, they are passed through `__` later.

File: lib/modules/apostrophe-login/index.js (hunk 2 above)
Comment (boutell) [reply lib/modules/apostrophe-login/index.js]: That... would take less logic. Will do.

**Comment by boutell** [comment]
Implemented your suggestion, merging.

**Pull request merged**